crates/sui-framework/packages/sui-framework/sources/collectible/witness_policy.move

```diff
@@ -0,0 +1,107 @@
+/// to receive a matching receipt.
+///
+/// One important use case for this policy is the ability to lock something
+/// in the `Kiosk`. When an item is placed into the Kiosk, a `PlacedWitness`
+/// struct is created which can be used to prove that the `T` was placed
+/// to the `Kiosk`.
+module sui::witness_policy {
+    use sui::transfer_policy::{
+        Self as policy,
+        TransferPolicy,
+        TransferPolicyCap,
+        TransferRequest
+    };
+
+    /// When a Proof does not find its Rule<Proof>.
+    const ERuleNotFound: u64 = 0;
+
+    /// Custom witness-key for the "proof policy".
+    struct Rule<phantom Proof: drop> has drop {}
+
+    /// Creator action: adds the Rule.
+    /// Requires a "Proof" witness confirmation on every transfer.
+    public fun set<T: key + store, Proof: drop>(
+        policy: &mut TransferPolicy<T>,
+        cap: &TransferPolicyCap<T>
+    ) {
+        policy::add_rule(Rule<Proof> {}, policy, cap, true);
+    }
+
+    /// Buyer action: follow the policy.
+    /// Present the required "Proof" instance to get a receipt.
+    public fun prove<T: key + store, Proof: drop>(
+        _proof: Proof,
+        policy: &TransferPolicy<T>,
+        request: &mut TransferRequest<T>
+    ) {
+        assert!(policy::has_rule<T, Rule<Proof>>(policy), ERuleNotFound);
+        policy::add_receipt(Rule<Proof> {}, request)
+    }
+}
+
+#[test_only]
+module sui::witness_policy_tests {
+    use sui::witness_policy;
+    use sui::tx_context::dummy as ctx;
+    use sui::transfer_policy as policy;
+    use sui::transfer_policy_tests::{
+        Self as test,
+        Asset
+    };
+
+    /// Confirmation of an action to use in Policy.
+    struct Proof has drop {}
+
+    /// Malicious attempt to use a different proof.
+    struct Cheat has drop {}
+
+    #[test]
+    fun test_default_flow() {
+        let ctx = &mut ctx();
+        let (policy, cap) = test::prepare(ctx);
+
+        // set the lock policy and require `Proof` on every transfer.
+        witness_policy::set<Asset, Proof>(&mut policy, &cap);
+
+        let request = policy::new_request(0, test::fresh_id(ctx), ctx);
+
+        witness_policy::prove(Proof {}, &policy, &mut request);
+        policy::confirm_request(&policy, request);
+        test::wrapup(policy, cap, ctx);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = sui::transfer_policy::EPolicyNotSatisfied)]
+    fun test_no_proof() {
+        let ctx = &mut ctx();
+        let (policy, cap) = test::prepare(ctx);
+
+        // set the lock policy and require `Proof` on every transfer.
+        witness_policy::set<Asset, Proof>(&mut policy, &cap);
+        let request = policy::new_request(0, test::fresh_id(ctx), ctx);
+
+        policy::confirm_request(&policy, request);
+        test::wrapup(policy, cap, ctx);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = sui::witness_policy::ERuleNotFound)]
+    fun test_wrong_proof() {
```

Comment:
Is this policy supposed to be used with the witnesses that come from `Kiosk`?  If so I still don't quite see how we can prevent someone from creating a witness from actions on one instance and then using it as proof on another witness.

Reply:
Good question. 

Given that TransferPolicy<T> affects all items of a type T, and kiosk witnesses are also returned "typed", only an item of type T can result in KioskWitness<T>. To cheat on this system, one has to put / list another item T. And that is okay.

While implementing I considered two main applications:

1. TransferPolicy has "allow_taking" set to "false". Once placed into the Kiosk, items can't be taken. The only thing that can happen to them is list-sell (simple) or PurchaseCap-grab (more complex). However, we can't lock items in the `purchase` function - its signature is `purchase() :(T, TransferRequest<T>)`. Hence we need to find a way to "force" user to put it into kiosk - and the "PlacedWitness" in the Kiosk allows us to achieve that.

2. We issue a "key-only" type, eg `Collectible has key`. To sell the Collectible, we need to turn it into a transferable type "TradableCollectible" - but we need a guarantee that it was listed / placed into the Kiosk. PlacedWitness or ListedWitness again help. Once sold, we can issue a rule for a TransferPolicy which will force the `TradableCollectible` to be turned into non-transferable Collectible again.